src/validation.cpp

```diff
@@ -2566,14 +2566,14 @@ bool FindUndoPos(CValidationState& state, int nFile, CDiskBlockPos& pos, unsigne
     return true;
 }
 
-bool CheckColdStakeFreeOutput(const CTransaction& tx, const int nHeight)
+static bool CheckColdStakeFreeOutput(const CTransactionRef& tx, const int nHeight)
 {
-    if (!tx.HasP2CSOutputs())
+    if (!tx->HasP2CSOutputs())
```

Comment:
We could add a minor performance upgrade here.
We don't really need to check all outputs (which could be many, e.g. for a regular coinstake splitting a big utxo), we can just check if `tx->vout[1].scriptPubKey.IsPayToColdStaking()`. 
The script-check ensures that, if this is a valid coinstake, then the scriptPubKey of the outs 1...(n-1) are equal.